File: Prediction.m
classdef Prediction
    methods (Static)
        function [xhatNow,estPreErrNowNorm,bestLevel]=AMLP(xopt,maxLevel)
            % Inputs:
                % xopt:
                    % The time history of a particular global optimum (or its 
                    % approximation) at different time steps (time steps 0, 1, ..., t). 
                    % It is a matrix of size N*D, in which D is the problem dimension 
                    % and N is the number of solutions. Only the last maxLevel rows are
                    % required, so N can be less than t. However, this code 
                    % automatically discard older solution if they are provided. 
                % maxLevel:
                    % the maximum level considered for prediction (e.g. 20)
            % Outputs:
                % xhat:
                    % the predicted solution (global optimum) by each prediction level 
                    % in time step t+1 (currentTimeStep)
                % estPreErr:
                    % Estimated prediction error for time step t+1 from each prediction
                    % level
                % bestLevel:
                    % The most reliable prediction level for time step t+1, which can
                    % be used to define the center and variation of the seed 
                    % population for time step t+1

            [T,D]=size(xopt); 
            %% Calculate the prediction and the prediction error from all requested and possible levels
            [xhat,preErr,maxLevelPrime]=Prediction.MLP(xopt,maxLevel); % see MLP.m for explanation of the inputs/outputs
            estPreErrNowNorm=nan(1,maxLevelPrime); % preallocation of the norm of the estimated prediction error for the prediction in the new time step
            xhatNow=nan(maxLevelPrime,D); % preallocation of the predicted optimum for the new time step 
            %% store the prediction and the estimated error from each level
            for L=1:maxLevelPrime
                estPreErrNowNorm(L)=norm(preErr{L}(end,:)); 
                xhatNow(L,:)=xhat{L}(end,:);
            end
            %% return the outcome of the best prediction level
            if T==1 % no other prediction can be made for the second time step
                bestLevel=1;
                estPreErrNowNorm=nan; % No estimation for the prediction error can be calculated
            else
                [~,bestLevel]=min(estPreErrNowNorm);
                % now if the following two conditions are satisifed, consider the next level as the best level 
                if (bestLevel==maxLevelPrime-1) && (maxLevelPrime==T)
                    bestLevel=bestLevel+1;
                end
            end
        end
 


        function [xhat,preErr,maxLevelPrime]=MLP(xopt,maxLevel)
            % Inputs:
                % xopt:
                    % The time history of a particular global optimum at time steps 1, ..., T 
                    % It is a matrix of size T*D, in whiche maximum prediction level is
                    % automatically set to T.  
            % Outputs:
                % xhat:
                    % the predicted solution (global optimum) by each prediction level 
                    % for each level and each time step   
                % preErr:
                    % prediction error for each level and each time step

            [t,D]=size(xopt); %  the number of solution and the problem dimension.  
            t=t-1; % the just finished time step 

            maxLevelPrime=min(maxLevel,t+1); % you cannot/should not calculate prediction levels higher than this
            t_recent=max(0,t-maxLevel); % solutions before this time steps are not used for prediction

            xhat=cell(1,maxLevelPrime); % predicted global optimum (preallocation) 
            preErr=cell(1,maxLevelPrime);  % prediction error (preallocation)

            %% calculate the prediction solution and the corresponding prediction error for each time step at level 1
            L=1; % prediction level
            xhat{L}=nan(t+2,D); % preallocation for prediction solution
            preErr{L}=nan(t+1,D);
            for tau=t_recent:t
                xhat{L}(tau+2,:)=xopt(tau+1,:); % Predicted solution for time step tau
                if tau<t
                    preErr{L}(tau+2,:)=xopt(tau+2,:)-xhat{L}(tau+2,:);
                end
            end

            %% calculate predicted solution and predicted error for higher levels  
            for L=2:maxLevelPrime
                xhat{L}=nan(t+2,D); % preallocation for prediction solution
                preErr{L}=nan(t+1,D); % preallocation for prediction solution

                for tau=t_recent:t
                    if tau>=L-1
                        xhat{L}(tau+2,:)=xhat{L-1}(tau+2,:)+preErr{L-1}(tau+1,:);
                        if tau<t
                            preErr{L}(tau+2,:)=xopt(tau+2,:)-xhat{L}(tau+2,:);
                        end
                    end
                end
            end 
            % The following lines deals with an exceptional case, in which the 
            % predicted solution from the Level maxLevelPrime is available but no 
            % estimate for this prediction is available. In this case, the value of the
            % lower level is used
            if (maxLevelPrime==t+1) && (maxLevelPrime>1) 
                preErr{maxLevelPrime}(t+1,:)=preErr{maxLevelPrime-1}(t+1,:);
            end
        end
    end
end

File: prediction.py
"""Prediction methods for dynamic optimization (AMLP)."""
import numpy as np
from typing import Tuple, List


class Prediction:
    """Adaptive multilevel prediction methods."""
    
    @staticmethod
    def AMLP(xopt: np.ndarray, maxLevel: int) -> Tuple[np.ndarray, np.ndarray, int]:
        """Adaptive Multi-Level Prediction.
        
        Parameters
        ----------
        xopt : (T, D) array
            Time history of a particular global optimum at different time steps.
        maxLevel : int
            Maximum level considered for prediction.
            
        Returns
        -------
        xhatNow : (maxLevelPrime, D) array
            Predicted solution from each prediction level for time step t+1.
        estPreErrNowNorm : (maxLevelPrime,) array
            Estimated prediction error norm for each level.
        bestLevel : int
            Most reliable prediction level for time step t+1.
        """
        T, D = xopt.shape
        
        # Calculate the prediction and error from all requested levels
        xhat, preErr, maxLevelPrime = Prediction.MLP(xopt, maxLevel)
        
        estPreErrNowNorm = np.full(maxLevelPrime, np.nan)
        xhatNow = np.full((maxLevelPrime, D), np.nan)
        
        for L in range(maxLevelPrime):
            estPreErrNowNorm[L] = np.linalg.norm(preErr[L][-1, :])
            xhatNow[L, :] = xhat[L][-1, :]
        
        # Return the outcome of the best prediction level
        if T == 1:
            bestLevel = 1  # 1-based index as in MATLAB
            estPreErrNowNorm = np.array([np.nan])
        else:
            bestLevel = int(np.argmin(estPreErrNowNorm)) + 1  # 1-based
            # If best is second-to-last and maxLevelPrime == T, use next level
            if (bestLevel == maxLevelPrime - 1) and (maxLevelPrime == T):
                bestLevel = bestLevel + 1
        
        return xhatNow, estPreErrNowNorm, bestLevel
    
    @staticmethod
    def MLP(xopt: np.ndarray, maxLevel: int) -> Tuple[List[np.ndarray], List[np.ndarray], int]:
        """Multi-Level Prediction.
        
        Parameters
        ----------
        xopt : (T, D) array
            Time history of a particular global optimum.
        maxLevel : int
            Maximum prediction level.
            
        Returns
        -------
        xhat : List of (t+2, D) arrays
            Predicted solution by each prediction level for each time step.
        preErr : List of (t+1, D) arrays
            Prediction error for each level and each time step.
        maxLevelPrime : int
            Actual maximum level used.
        """
        t, D = xopt.shape
        t = t - 1  # Last finished time step (0-based)
        
        maxLevelPrime = min(maxLevel, t + 1)
        t_recent = max(0, t - maxLevel)
        
        xhat = [None] * maxLevelPrime
        preErr = [None] * maxLevelPrime
        
        # Level 1
        L = 0  # 0-based indexing in Python
        xhat[L] = np.full((t + 2, D), np.nan)
        preErr[L] = np.full((t + 1, D), np.nan)
        for tau in range(t_recent, t + 1):
            xhat[L][tau + 1, :] = xopt[tau, :]  # tau+1 in MATLAB becomes tau in 0-based
            if tau < t:
                preErr[L][tau + 1, :] = xopt[tau + 1, :] - xhat[L][tau + 1, :]
        
        # Higher levels
        for L in range(1, maxLevelPrime):
            xhat[L] = np.full((t + 2, D), np.nan)
            preErr[L] = np.full((t + 1, D), np.nan)
            for tau in range(t_recent, t + 1):
                if tau >= L:  # L is prediction level (1-based originally, 0-based here)
                    xhat[L][tau + 1, :] = xhat[L - 1][tau + 1, :] + preErr[L - 1][tau, :]
                    if tau < t:
                        preErr[L][tau + 1, :] = xopt[tau + 1, :] - xhat[L][tau + 1, :]
        
        # Handle exceptional case
        if (maxLevelPrime == t + 1) and (maxLevelPrime > 1):
            preErr[maxLevelPrime - 1][t, :] = preErr[maxLevelPrime - 2][t, :]
        
        return xhat, preErr, maxLevelPrime
